lld/ELF/InputFiles.cpp

```diff
@@ -918,6 +918,60 @@ void ObjFile<ELFT>::initializeSections(bool ignoreComdats,
     handleSectionGroup<ELFT>(this->sections, entries);
 }
 
+template <typename ELFT>
+static void parseGnuPropertyNote(Ctx &ctx, uint32_t &featureAndType,
+                                 ArrayRef<uint8_t> &desc, ELFFileBase *f,
+                                 const uint8_t *base,
+                                 ArrayRef<uint8_t> *data = nullptr,
+                                 StringRef sectionName = ".note.gnu.property") {
+  auto err = [&](const uint8_t *place) -> ELFSyncStream {
+    auto diag = Err(ctx);
+    diag << f->getName() << ":(" << sectionName << "+0x"
+         << Twine::utohexstr(place - base) << "): ";
+    return diag;
+  };
+
+  while (!desc.empty()) {
+    const uint8_t *place = desc.data();
+    if (desc.size() < 8)
+      return void(err(place) << "program property is too short");
+    uint32_t type = read32<ELFT::Endianness>(desc.data());
+    uint32_t size = read32<ELFT::Endianness>(desc.data() + 4);
+    desc = desc.slice(8);
+    if (desc.size() < size)
+      return void(err(place) << "program property is too short");
+
+    if (type == featureAndType) {
+      // We found a FEATURE_1_AND field. There may be more than one of these
+      // in a .note.gnu.property section, for a relocatable object we
+      // accumulate the bits set.
+      if (size < 4)
+        return void(err(place) << "FEATURE_1_AND entry is too short");
+      f->andFeatures |= read32<ELFT::Endianness>(desc.data());
+    } else if (ctx.arg.emachine == EM_AARCH64 &&
+               type == GNU_PROPERTY_AARCH64_FEATURE_PAUTH) {
+      // If the file being parsed is a SharedFile, we cannot pass in
+      // the data variable as there is no InputSection to collect the
+      // data from. As such, these are ignored. They are needed either
+      // when loading a shared library oject.
```

Comment:
IIUC the data->data() is the full contents of the notes section. Whereas desc->data() is the contents of just one desc. Instead of skipping both the errors you could do something like
```
ArrayRef<uint8_t> contents = data ? *data : desc;
```
Then use contents.data() below.

Reply:
The comment is a bit too verbose regarding the lld code convention.

Actually, I am not sure shared objects should parse GNU_PROPERTY_AARCH64_FEATURE_PAUTH. It's orthogonal to this PR and this PR should not introduce the behavior change